src/holeyvec.rs

```diff
@@ -0,0 +1,366 @@
+    ///
+    /// v.push(1);
+    /// v.push(2);
+    /// v.push(3);
+    /// v.remove(1);
+    /// let mut it = v.iter();
+    /// assert_eq!(it.next(), Some(&1));
+    /// assert_eq!(it.next(), Some(&3));
+    /// ```
+    pub fn iter(&self) -> Iter<T> {
+        Iter::new(self)
+    }
+
+    /// Return mutable iterator over non empty elements of the vector.
+    /// 
+    /// ## Examples
+    /// ```
+    /// use holeyvec::HoleyVec;
+    ///
+    /// let mut v = HoleyVec::new();
+    /// let mut it = v.iter_mut();
+    /// assert_eq!(it.next(), None);
+    ///
+    /// v.push(1);
+    /// v.push(2);
+    /// v.push(3);
+    /// v.remove(1);
+    /// let mut it = v.iter_mut();
+    /// assert_eq!(it.next(), Some(&mut 1));
+    /// assert_eq!(it.next(), Some(&mut 3));
+    /// ```
+    pub fn iter_mut(&mut self) -> IterMut<T> {
+        IterMut::new(self)
+    }
+}
+
+impl<T> std::ops::Index<usize> for HoleyVec<T> {
+    type Output = T;
+
+    fn index(&self, index: usize) -> &Self::Output {
+        self.get(index).expect("Index doesn't exist")
+    }
+}
+
+impl<T> std::ops::IndexMut<usize> for HoleyVec<T> {
+
+    fn index_mut(&mut self, index: usize) -> &mut Self::Output {
+        self.get_mut(index).expect("Index doesn't exist")
+    }
+}
+
+/// [HoleyVec] iterator to iterate through non-empty elements of the vector.
+pub struct Iter<'a, T> {
+    delegate: std::slice::Iter<'a, Cell<T>>
+}
+
+impl<'a, T> Iter<'a, T> {
+    fn new(vec: &'a HoleyVec<T>) -> Self {
+        Self{ delegate: vec.vec.iter() }
+    }
+}
+
+impl<'a, T> std::iter::Iterator for Iter<'a, T> {
+    type Item = &'a T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            match self.delegate.next() {
+                None => return None,
+                Some(Cell::Hole(_)) => continue,
+                Some(Cell::Value(value)) => return Some(value),
+            }
+        }
+    }
+}
+
+impl<'a, T> IntoIterator for &'a HoleyVec<T> {
+    type Item = &'a T;
+    type IntoIter = Iter<'a, T>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.iter()
+    }
+}
+
+/// [HoleyVec] iterator to iterate through non-empty elements of the vector.
+pub struct IterMut<'a, T> {
+    delegate: std::slice::IterMut<'a, Cell<T>>
+}
+
+impl<'a, T> IterMut<'a, T> {
+    fn new(vec: &'a mut HoleyVec<T>) -> Self {
+        Self{ delegate: vec.vec.iter_mut() }
+    }
+}
+
+impl<'a, T> std::iter::Iterator for IterMut<'a, T> {
+    type Item = &'a mut T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            match self.delegate.next() {
+                None => return None,
+                Some(Cell::Hole(_)) => continue,
+                Some(Cell::Value(value)) => return Some(value),
+            }
+        }
+    }
+}
+
+impl<'a, T> IntoIterator for &'a mut HoleyVec<T> {
+    type Item = &'a mut T;
+    type IntoIter = IterMut<'a, T>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.iter_mut()
+    }
+}
```

Comment:
Here is another convenience implementation so you can collect into a HoleyVec:
```rust
impl<T> FromIterator<T> for HoleyVec<T> {
    fn from_iter<I: IntoIterator<Item=T>>(iter: I) -> Self {
        let vec: Vec<Cell<T>> = iter.into_iter().map(|t| Cell::Value(t)).collect();
        Self {first_hole: vec.len(), vec }
    }
}
```

Reply:
Thank you! Added in https://github.com/vsbogd/holeyvec/pull/1/commits/b2771f367692e92fca631195c3d3ffaa625365f5